File: prob2.py
# Time Complexity : O(n**2)
# Space Complexity :O(n)
# Passed on Leetcode: yes

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)

        dp = [i for i in triangle[n-1]]

        for i in range(n-2, -1, -1):
            for j in range(len(triangle[i])):
                dp[j] = triangle[i][j] + min(dp[j], dp[j+1])


        return dp[0]

File: prob2.java
// Time Complexity : O(n^2)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes


class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        List<int[]> ways = new ArrayList<>();
        int capacity = 3;
        int[] tmp;
        ways.add(new int[]{Integer.MAX_VALUE, triangle.get(0).get(0),Integer.MAX_VALUE});
        for (int i = 1; i < triangle.size(); i++)
        {
            capacity++;
            tmp = new int[capacity];
            tmp[0] = Integer.MAX_VALUE;
            tmp[capacity - 1] = Integer.MAX_VALUE;
            for (int j = 1; j < capacity - 1; j++)
                tmp[j] = (Math.min(ways.get(i - 1)[j - 1], ways.get(i - 1)[j]) + triangle.get(i).get(j - 1));
            ways.add(tmp);
        }
        int min = ways.get(ways.size() - 1)[1];
        for (Integer num :ways.get(ways.size() - 1))
            min = Math.min(min, num);
        return min;
    }
}